File: app/src/main/java/ds/photosight/ui/widget/Histogram.java
package ds.photosight.ui.widget;

import android.animation.ValueAnimator;
import android.content.Context;
import android.graphics.*;
import android.util.AttributeSet;
import android.view.View;
import com.androidquery.util.AQUtility;
import ds.photosight.R;
import ds.photosight.utils.L;

public class Histogram extends View {

	private int[] data;
	private Bitmap bitmap;
	private Bitmap result;
	private Paint paint;
	private ValueAnimator mAnimator;
	private int alpha = 0;


	public Histogram(final Context context, final AttributeSet attrs) {
		super(context, attrs);

		paint = new Paint(Paint.ANTI_ALIAS_FLAG);
		paint.setColor(Color.GRAY);
		paint.setTextSize(AQUtility.dip2pixel(context, 14));

		fillDemo();
	}


	private void fillDemo() {


	}


	public void setData(int[] data) {
		this.data = data;
	}


	public void setBitmap(Bitmap b) {
		bitmap = b;
		invalidate();
	}


	@Override
	protected void onSizeChanged(final int w, final int h, final int oldw, final int oldh) {
		super.onSizeChanged(w, h, oldw, oldh);
		L.v(String.format("w=%s h=%s", w, h));
		if (bitmap != null)
			process();
	}


	@Override
	protected void onDraw(final Canvas canvas) {
		int w = getWidth();
		int h = getHeight();

		if (bitmap != null) {
			if (result == null) {
				// process
				process();
			} else {
				// draw
				if (mAnimator == null && alpha == 0)
					startAnimator();

				paint.setAlpha(alpha);
				canvas.drawBitmap(result, 0, 0, paint);
			}

		}

		String text = getContext().getString(R.string.processing);
		paint.setAlpha(255 - alpha);
		canvas.drawText(text, (w - paint.measureText(text)) / 2, h / 2, paint);

	}


	private void startAnimator() {
		mAnimator = ValueAnimator.ofInt(0, 255);
		mAnimator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
			@Override
			public void onAnimationUpdate(final ValueAnimator animation) {
				alpha = (int) animation.getAnimatedValue();
				postInvalidate();
			}
		});
		mAnimator.start();

	}


	public void process() {
		AQUtility.postAsync(new Runnable() {
			@Override
			public void run() {

				L.v("process...");
				int[] r = new int[256];
				int[] g = new int[256];
				int[] b = new int[256];
				final int[] pixels = new int[bitmap.getWidth() * bitmap.getHeight()];
				bitmap.getPixels(pixels, 0, bitmap.getWidth(), 0, 0, bitmap.getWidth(), bitmap.getHeight());

				for (int p : pixels) {
					r[Color.red(p)]++;
					g[Color.green(p)]++;
					b[Color.blue(p)]++;
				}

				int max = Math.max(Math.max(max(r), max(g)), max(b));
				int height = getHeight();
				int width = getWidth();
				float minFactor = height / 12000f;
				float xFactor = width / 256f;
				float yFactor = Math.max((float) height / max, minFactor);
				L.v("xFactor=" + xFactor + " yFactor=" + yFactor);

				Paint paint = new Paint(Paint.ANTI_ALIAS_FLAG);
				paint.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.SCREEN));
				Path pathR = new Path();
				Path pathG = new Path();
				Path pathB = new Path();
				result = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);
				Canvas c = new Canvas(result);

				pathR.moveTo(0, height);
				pathG.moveTo(0, height);
				pathB.moveTo(0, height);

				for (int i = 0; i < 255; i++) {

					float x1 = i * xFactor;
					float y1 = height - r[i] * yFactor;
					float x2 = (i + 1) * xFactor;
					float y2 = height - r[i + 1] * yFactor;

					pathR.quadTo(x1, y1, (x1 + x2) / 2, (y1 + y2) / 2);

					y1 = height - g[i] * yFactor;
					y2 = height - g[i + 1] * yFactor;
					pathG.quadTo(x1, y1, (x1 + x2) / 2, (y1 + y2) / 2);

					y1 = height - b[i] * yFactor;
					y2 = height - b[i + 1] * yFactor;
					pathB.quadTo(x1, y1, (x1 + x2) / 2, (y1 + y2) / 2);

				}
				pathR.lineTo(width, height);
				pathG.lineTo(width, height);
				pathB.lineTo(width, height);
				paint.setColor(Color.RED);
				c.drawPath(pathR, paint);
				paint.setColor(Color.GREEN);
				c.drawPath(pathG, paint);
				paint.setColor(Color.BLUE);
				c.drawPath(pathB, paint);

				postInvalidate();
			}
		});
	}


	private static int max(int[] values) {
		int max = Integer.MIN_VALUE;
		for (int value : values) {
			if (value > max)
				max = value;
		}
		return max;
	}

}

File: app/src/main/java/ds/photosight/ui/widget/Histogram.kt
package ds.photosight.ui.widget

import android.animation.ValueAnimator
import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.view.View
import ds.photosight.utils.dp
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import timber.log.Timber
import kotlin.math.max

class Histogram(context: Context, attrs: AttributeSet) : View(context, attrs) {

    private var bitmap: Bitmap? = null
    private var result: Bitmap? = null
    private val paint: Paint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        color = Color.GRAY
        textSize = 14.dp.toFloat()
    }
    private var animator: ValueAnimator? = null
    private var alpha = 0

    fun setBitmap(b: Bitmap) {
        bitmap = b
        invalidate()
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        bitmap?.let { process(it) }
    }

    override fun onDraw(canvas: Canvas) {
        if (bitmap != null) {
            if (result == null) {
                process(bitmap!!)
            } else {
                if (animator == null && alpha == 0) {
                    startAnimator()
                }

                paint.alpha = alpha
                canvas.drawBitmap(result!!, 0f, 0f, paint)
            }

        }

    }


    private fun startAnimator() {
        animator = ValueAnimator.ofInt(0, 255)
        animator!!.addUpdateListener {
            alpha = it.animatedValue as Int
            postInvalidate()
        }
        animator!!.start()

    }


    private fun process(bitmap: Bitmap) = GlobalScope.launch {
        Timber.d("process...")
        val r = IntArray(256)
        val g = IntArray(256)
        val b = IntArray(256)
        val pixels = IntArray(bitmap.width * bitmap.height)
        bitmap.getPixels(pixels, 0, bitmap.width, 0, 0, bitmap.width, bitmap.height)

        for (p in pixels) {
            r[Color.red(p)]++
            g[Color.green(p)]++
            b[Color.blue(p)]++
        }

        val max = listOf(r.max()!!, g.max()!!, b.max()!!).max()!!
        val height = height
        val width = width
        val minFactor = height.toFloat() / 12000.toFloat()
        val xFactor = width.toFloat() / 256.toFloat()
        val yFactor = max(height.toFloat() / max.toFloat(), minFactor)
        Timber.d("xFactor=$xFactor yFactor=$yFactor")

        val paint = Paint(Paint.ANTI_ALIAS_FLAG)
        paint.xfermode = PorterDuffXfermode(PorterDuff.Mode.SCREEN)
        val pathR = Path()
        val pathG = Path()
        val pathB = Path()
        result = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
        val c = Canvas(result!!)

        pathR.moveTo(0f, height.toFloat())
        pathG.moveTo(0f, height.toFloat())
        pathB.moveTo(0f, height.toFloat())

        for (i in 0 until 255) {

            val x1 = i.toFloat() * xFactor
            var y1 = height.toFloat() - r[i].toFloat() * yFactor
            val x2 = (i + 1).toFloat() * xFactor
            var y2 = height.toFloat() - r[i + 1].toFloat() * yFactor

            pathR.quadTo(x1, y1, (x1 + x2) / 2, (y1 + y2) / 2)

            y1 = height.toFloat() - g[i].toFloat() * yFactor
            y2 = height.toFloat() - g[i + 1].toFloat() * yFactor
            pathG.quadTo(x1, y1, (x1 + x2) / 2, (y1 + y2) / 2)

            y1 = height.toFloat() - b[i].toFloat() * yFactor
            y2 = height.toFloat() - b[i + 1].toFloat() * yFactor
            pathB.quadTo(x1, y1, (x1 + x2) / 2, (y1 + y2) / 2)

        }
        pathR.lineTo(width.toFloat(), height.toFloat())
        pathG.lineTo(width.toFloat(), height.toFloat())
        pathB.lineTo(width.toFloat(), height.toFloat())
        paint.color = Color.RED
        c.drawPath(pathR, paint)
        paint.color = Color.GREEN
        c.drawPath(pathG, paint)
        paint.color = Color.BLUE
        c.drawPath(pathB, paint)

        postInvalidate()
    }

}
